Improve error message for impossible JoinKinds constraint

adamgundry opened the pull request:
Fixes #423.

WARNING: typeclass Prolog ahead.

adamgundry committed:
Improve error message for impossible JoinKinds constraint (fixes #423)

arybczak commented:
Thanks. Do we want to drop support for GHC 8.2 and 8.4 (as this introduces `UndecidableSuperClasses`)? I personally don't mind.

Also, doctests fail, presumably due to a missing import.

phadej commented on optics-core/src/Optics/Internal/Optic/Subtyping.hs:
```diff
@@ -432,8 +444,80 @@ instance JoinKinds A_Setter           A_Traversal        A_Setter           wher
+-- @is_concrete@ will be unified with @False@.
+--
+class IsOpticKind (m :: OpticKind) (is_concrete :: Bool) | m -> is_concrete
+instance is_concrete ~ 'True => IsOpticKind An_AffineFold      is_concrete
```
What will happen if a kind is forgotten, we get a "wrong" custom type error?

adamgundry replied:
Yes, I was a bit dissatisfied with the need to list all the optic kinds explicitly, but I couldn't immediately see how to avoid it. Having thought a bit more I've got a version that doesn't need this.

adamgundry committed:
Replace IsOpticKind with IsConcrete

This is less vulnerable to us forgetting to extend the definition if
we add a new optic kind.

adamgundry commented:
I was playing around with this branch and noticed this:
```
> view @A_Getter (folded % _) [Just "abc", Nothing, Just "def"]

<interactive>:3:17: error:
    • A_Fold composed with l0 produces p0,
      but the context expects A_Getter
```
We should be able to do better here, by checking if more parameters are concrete.

On the other hand, we're a bit stuck if the context doesn't force a particular optic kind (cf. #308):
```
> view (folded % _) [Just "abc", Nothing, Just "def"]

<interactive>:4:1: error:
    • Overlapping instances for Is k0 A_Getter
        arising from a use of ‘view’
...
    • Overlapping instances for JoinKinds A_Fold l0 k0
        arising from a use of ‘%’
```

arybczak commented:
What's the status of this PR?

adamgundry commented:
I've been away on holiday so apologies for the radio silence. I'm still planning to polish this a bit, but it may be a little while before I can get back to it.

arybczak commented:
@adamgundry I posted an alternative solution in #439, please have a look if you have a moment. It's not as fancy, but should be more predictable :)

arybczak commented:
Closing since #439 was merged a long time ago and it more or less supersedes it.
